src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -305,6 +305,59 @@ private async Task ReplyOnReAuthenticationAsync<TIOAdapter>(TIOAdapter adapter,
             }
         }
 
+        // This will initiate renegotiation or PHA for Tls1.3
+        private async Task RenegotiateAsync(CancellationToken cancellationToken)
+        {
+            if (Interlocked.Exchange(ref _nestedAuth, 1) == 1)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "NegotiateClientCertificateAsync", "renegotiate"));
+            }
+
+            if (Interlocked.CompareExchange(ref _nestedRead, 1, 0) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
+            }
+
+            if (Interlocked.CompareExchange(ref _nestedWrite, 1, 0) == 1)
+            {
+                _nestedRead = 0;
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(WriteAsync), "write"));
+            }
+
+            _sslAuthenticationOptions!.RemoteCertRequired = true;
+            IReadWriteAdapter adapter = new AsyncReadWriteAdapter(InnerStream, cancellationToken);
+
+            try
+            {
+                SecurityStatusPal status = _context!.Renegotiate(out byte[]? nextmsg);
+                if (nextmsg?.Length > 0)
+                {
+                    await adapter.WriteAsync(nextmsg, 0, nextmsg.Length).ConfigureAwait(false);
+                    await adapter.FlushAsync().ConfigureAwait(false);
+                }
+
+                if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
```

Comment:
Should we check this before we check for nextmsg above? Would it ever be the case that thee status is not OK, but we still want to write something to the peer? How does that work?

Reply:
I don't know this would happen in practice but for handshake the OS can generate TlsAlert for failed cases. In that case we want to send it to the peer before failing locally. This replicates the structure from `ForceAuthenticationAsync` but we can change that if we want to.

Reply:
If it's mimicking what we do in other cases then that seems fine.